src/PrettyPrint.hs

```diff
@@ -0,0 +1,224 @@
+            <> "(" 
+            <> T.concat (map ((<> ", ") . prettyPrint 0) exprs) 
+            <> ")"
+    prettyPrint i (Array exprs _ _) =
+        indent i 
+            <> "[<" 
+            <> T.concat (map ((<> ", ") . prettyPrint 0) exprs) 
+            <> ">]"
+    prettyPrint i (App f arg a) =
+        indent i <> "(" <> prettyPrint 0 f <> " " <> prettyPrint 0 arg <> ")"
+    prettyPrint i (Match expr cases _ _) =
+        indent i <> "match " <> prettyPrint 0 expr <> " with\n"
+                 <> T.concat (map ((<> "\n") . prettyPrint (i + 1)) cases)
+    prettyPrint i (External path id _ _) = 
+        indent i <> "(_external " <> id <> " " <> id <> ")"
+    prettyPrint i (Internal (Identifier id) _ _) =
+        indent i <> "(_internal " <> id <> ")"
+    prettyPrint i (AnnotatedExpr expr sig _ _) =
+        indent i <> prettyPrint 0 expr <> " : " <> prettyPrint 0 sig
+    prettyPrint i (FormatString fmts _ _) =
+        indent i <> "\"" <> T.concat (map (prettyPrint 0) fmts) <> "\""
+
+instance PrettyPrint (MatchCase a) where
+    prettyPrint i (MatchCase pat expr) =
+        indent i 
+            <> "case " <> prettyPrint 0 pat
+            <> " => "  <> prettyPrint 0 expr
+
+instance PrettyPrint (FormatExpr a) where
+    prettyPrint i (FmtStr str) = indent i <> str
+    prettyPrint i (FmtExpr expr) = 
+        indent i <> "{" <> prettyPrint 0 expr <> "}"
+
+instance PrettyPrint PrimExpr where
+    prettyPrint i (IntLit int)  = indent i <> T.pack (show int)
+    prettyPrint i (CharLit c)   = indent i <> "'" <> T.cons c "'"
+    prettyPrint i (FloatLit f)  = indent i <> T.pack (show f)
+    prettyPrint i (StringLit s) = indent i <> "\"" <> s <> "\""
+    prettyPrint i (BoolLit b)   = indent i <> T.pack (show b)
+    prettyPrint i UnitLit       = indent i <> "()"
+
+instance PrettyPrint TypeSig where
+    prettyPrint i (TypeSig constraints typeSig) =
+        indent i 
+        <> "(" <> T.concat (map ((<>  ", ") . prettyPrint 0) constraints) <> ")"
+        <> " => " <> prettyPrint 0 typeSig
+
+instance PrettyPrint Constraint where
+    prettyPrint i (Constraint (TypeName name) (TypeVar param)) =
+        indent i <> name <> " " <> param
+
+instance PrettyPrint Type where
+    prettyPrint i (TypeVariable (TypeVar var)) = indent i <> var
+    prettyPrint i (PrimitiveType t) = indent i <> prettyPrint 0 t
+    prettyPrint i (FunctionType from to) = 
+        indent i 
+            <> "(" 
+            <> prettyPrint 0 from 
+            <> " -> " 
+            <> prettyPrint 0 to 
+            <> ")"
+    prettyPrint i (NonPrimType (TypeName t)) = indent i <> t
+    prettyPrint i (ParamType (TypeName t) args) =
+        indent i <> t <> T.concat (map ((" " <>) . prettyPrint 0) args)
+    prettyPrint i (PolymorphicParamType (TypeVar var) args) =
+        indent i <> var <> T.concat (map ((" " <>) . prettyPrint 0) args)
+    prettyPrint i (TupleType types) =
+        indent i <> T.concat (map ((<> ", ") . prettyPrint 0) types)

```

Comment:
Same as line 47 + the type is flattened:
`(False, 6, (False, 6, ), ) :: Bool, Int, Bool, Int, , `

the expression should be of type:
`(Bool, Int, (Bool, Int))`

```suggestion
    prettyPrint i (TupleType types) =
        indent i 
            <> "("
            <> T.concat (mapWithoutLast ((<> ", ") . prettyPrint 0) types)
            <> ")"
```